changelog/replacementDchar.dd

```diff
@@ -0,0 +1,42 @@
+# Change autodecode from throwing `UTFException` on invalid encodings to returning
+`std.utf.replacementDchar` `('\uFFFD')` instead.
+
+When `char` and `wchar` slices are iterated, they are autodecoded meaning the UTF-8 and
+UTF-16 code unit sequences are automatically consumed and decoded into `dchar` code points.
+If an invalid sequence is encountered, a
+$(LINK2 https://dlang.org/phobos/std_utf.html#UTFExceptions,`UTFException`)
+is raised.
+
+This change instead causes
+$(LINK2 https://dlang.org/phobos/std_utf.html#replacementDchar, `std.utf.replacementDchar`)
+to be the decoded result of an invalid sequence.
+
+If the old behavior is desired, the autodecoding iteration can be replaced with
+$(LINK2 https://dlang.org/phobos/std_utf.html#byDchar, `std.utf.byDchar()`)
+to decode into `dchar`s followed by
+$(LINK2 https://dlang.org/phobos/std_algorithm_iteration.html#.filter, `std.algorithm.filter()`)
+to detect the `replacementDchar` and throw the exception, as in the following example:
+
+---
+import std.utf;
+import std.algorithm;
+import std.stdio;
+
+void test(string s)
+{
+    s.byDchar()
+     .filter!((c)
+              {
+                  if (c == replacementDchar)
+                      throw new UTFException("invalid UTF-8 encoding");
+                  return true;
+              }) ()
+     .write();
```

Comment:
This is too complicated for a breaking change.
It needs to be simpler for an easy migration path:

```d
private struct ValidationRange(R)
{
    R r;
    this(R r)
    {
        this.r = r;
    }

    @property dchar front() @safe pure
    {
        import std.utf;

        assert(r.length, "Attempting to fetch the front of an empty array of " ~ R.stringof);
        size_t i = 0;
        return decode!(UseReplacementDchar.no)(r, i);
    }

    @property dchar back(T)(scope const(T)[] a) @safe pure
    {
        import std.utf;

        assert(r.length, "Attempting to fetch the back of an empty array of " ~ T.stringof);
        size_t i = r.length - strideBack(a, r.length);
        return decode!(UseReplacementDchar.no)(r, i);
    }

    alias r this;

}

auto validationRange(T)(scope const(T)[] a)
if (isAutodecodableString!(T[]) && !isAggregateType!(T[]))
{
    return ValidationRange!(typeof(a))(a);
}
```

https://run.dlang.io/is/MgKafm

Reply:
A simpler scheme is to just encapsulate the filter into its own range, validate(), i.e.:
```
s.byDchar().validate().write();
```
which would work the same whether or not autodecode is present, which is where we want to get to anyway.

Reply:
Sure, though that will be a bit slower (more comparisons + more ranges).
Anyhow, my point was that if we do such a breaking change the solution to revert back to the current behavior must be simple. When we removed `delete`, we had `__delete` as an alternative. Whatever it does, a solution here needs to be similarly easy to reach.

Reply:
Something like this:
```
import std.utf;
import std.algorithm;
import std.stdio;

void test(string s)
{
    s.byDchar()
     .validateDchar()
     .write();
}

void main()
{
    test("hello");
    test("\xFFbetty");
}

auto validateDchar(R)(R r)
{
     return filter!((c)
              {
                  if (c == replacementDchar)
                      throw new UTFException("invalid UTF-8 encoding");
                  return true;
              })(r);
}
```